### PR TITLE
1735 usage metrics web crawlers

### PR DESCRIPTION
**Pull request by Castronova**
- added a robots middleware layer that rejects page requests from non-human calls
- added userid to resource details report
- added a README file to the hs_tracking app

### CODE REVIEW

**Comment by mjstealey** [comment]
Run on Jenkins at: [Build #47: origin/1735-usage-metrics-web-crawlers](http://ci.hydroshare.org:8080/job/hydroshare-pull-requests/47/)

**Comment by pkdash** [comment]
@Castronova Is this ready for review?

**Comment by Castronova** [comment]
@pkdash yes, it's ready for review

File: hs_tracking/tests.py [hunk 1]
```diff
@@ -129,9 +129,9 @@ def test_history_empty(self):
         self.assertEqual(response.status_code, 200)
         reader = csv.reader(StringIO(response.content))
         rows = list(reader)
-
+        print rows
```
Comment (alexlemann) [inline hs_tracking/tests.py]: Left over from debugging?

File: hs_core/robots.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+# (wget http://www.robotstxt.org/db/all.txt, python robot_detection.py all.txt)
+
+class CrawlerBlocker:
+    def __init__(self):
+        self.whitelist = getattr(settings, "BOT_WHITELIST", -1)
+
+    def process_request(self, request):
+
+        user_agent = request.META.get('HTTP_USER_AGENT', None)
+        request.is_crawler = 0
```
Comment (alexlemann) [inline hs_core/robots.py]: There is inconsistent use of 0/1 and True/False. You should stick to using True and False here from what I can tell.

File: hs_core/robots.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+from django.conf import settings
+from django.http import HttpResponseForbidden
+
+
+# note: robot_detection needs to be updated periodically (possible inside the Dockerfile)
+# (wget http://www.robotstxt.org/db/all.txt, python robot_detection.py all.txt)
+
+class CrawlerBlocker:
+    def __init__(self):
+        self.whitelist = getattr(settings, "BOT_WHITELIST", -1)
```
Comment (alexlemann) [inline hs_core/robots.py]: Using an empty list rather than -1 for a default value here might make more sense.

**Comment by Castronova** [comment]
@alexlemann I agree with your comments.  After talking with @mjstealey this morning, I plan to remove portions of this code since there are better ways to provide/deny access to crawlers than what I have implemented.  I am going to scale back some of these changes so that they remain within the scope of usage tracking.

File: hs_core/views/__init__.py [hunk 4]
```diff
@@ -870,9 +870,14 @@ def update(self, group_to_update, request):
 def my_resources(request, page):
     resource_collection = get_my_resources_list(request)
     context = {'collection': resource_collection}
-    
-    return context
 
+    import sys
+    sys.path.append('/pycharm-debug')
+    import pydevd
+    pydevd.settrace('10.20.1.11', port=21000, suspend=False)
```
Comment (aphelionz) [inline hs_core/views/__init__.py]: Dont foret to take out debug stuff

File: hs_tracking/management/commands/stats.py [hunk 5]
```diff
@@ -156,17 +158,19 @@ def resources_details(self):
...
-            hour=0,
-            minute=0,
-            second=0,
-            microsecond=0
-        )
+             hour=0,
+             minute=0,
+             second=0,
+             microsecond=0
+         )
```
Comment (aphelionz) [inline hs_tracking/management/commands/stats.py]: Nitpicking here but the extra whitespace is weird here. Did the PEP8 test not pick this up?

Comment (Castronova) [inline hs_tracking/management/commands/stats.py]: No this wasn't picked up by flake

**Comment by aphelionz** [comment]
+1

**Pull request merged**